velox/docs/develop/hash-table.rst

```diff
@@ -0,0 +1,117 @@
+==========
+Hash table
+==========
+
+
+Layout
+------
+
+HashTable is implemented as an array of buckets. It is a linear data structure.
+Each bucket uses 128 bytes (2 * 64 = 2 cache lines) and contains 16 slots.
+Each HashTable entry occupies one slot. HashTable’s capacity is the total number of slots: total
+number of buckets * 16. HashTable’s capacity is always a power of 2.
+
+Each bucket contains 16 slots. Each slot consists of 2 pieces: a tag (7 bits) and a pointer (6 bytes).
+There are a total of 16 tags and 16 pointers in a bucket. These are stored tags first, followed by
+pointers. Each tag occupies 1 byte (only 7 bits are used). 16 tags occupy 16 bytes. Each pointer
+occupies 6 bytes. 16 pointers occupy 48 bytes. There are 16 bytes left unused at the end of the bucket.
+These are referred to as padding.
+
+.. image:: images/hash-table-layout.png
+    :align: center
+
+A hash table is never full. There are always some empty slots. Velox allows HashTable to fill up to ⅞ of
+capacity before resizing.
+On resize HashTable’s capacity doubles.
+
+Individual buckets may be completely empty, partially filled or full. Buckets are filled left to right.
+If a bucket is partially full, then first N tags and N pointers are filled and the rest are free (N < 16).
+
+
+Inserting an entry
+------------------
+
+To insert a new entry we need to figure out which slot to put it in.
+A slot is identified by bucket and offset within the bucket. First, we compute a hash of the entry.
+Then, we compute a tag and a bucket number from the hash.
+
+We use 7 bits of the hash for the tag: bits 32-38 inclusive. We use N bits of the hash for the bucket
+starting from bit 8.
+
+Given hash number
```

Comment:
Looks like I forgot to clarify how N is computed and explain what is N for th example. Let's add this information.

```
We use 7 bits of the hash for the tag: bits 32-38 inclusive. We use N bits of the hash for the bucket
starting from bit 8.

Number of bits used for the bucket is decided based on hash table capacity. Remember that capacity is always a power of 2: 2^n. Each bucket stored 16 entries, hence, we need 2^(n-4) buckets to store 2^n entries. Hence, we need to use n-4 bits of the hash for the bucket.

Let's say we have a hash table that can store a million entries: 2^20 = 1,048,576. Here, n = 20 and N = n - 4 = 16. We will use 16 bits for the bucket.

Given hash number...
```